src/facade/dragonfly_connection.cc

```diff
@@ -100,6 +102,129 @@ void UpdateIoBufCapacity(const base::IoBuf& io_buf, ConnectionStats* stats,
+  bool Write(string_view blob);
+  bool Write(iovec* blobs, size_t len);
+};
+
+void TrafficLogger::ResetLocked() {
+  if (log_file) {
+    log_file->Close();
+    log_file.reset();
+  }
+}
+
+// Returns true if Write succeeded, false if it failed and the recording should be aborted.
+bool TrafficLogger::Write(string_view blob) {
+  auto ec = log_file->Write(io::Buffer(blob));
+  if (ec) {
+    LOG(ERROR) << "Error writing to traffic log: " << ec;
+    ResetLocked();
+    return false;
+  }
+  return true;
+}
+
+bool TrafficLogger::Write(iovec* blobs, size_t len) {
+  auto ec = log_file->Write(blobs, len);
+  if (ec) {
+    LOG(ERROR) << "Error writing to traffic log: " << ec;
+    ResetLocked();
+    return false;
+  }
+  return true;
+}
+
+thread_local TrafficLogger tl_traffic_logger{};  // nullopt while disabled
+
+void OpenTrafficLogger(string_view base_path) {
+  unique_lock lk{tl_traffic_logger.mutex};
+  if (tl_traffic_logger.log_file)
+    return;
+
+  // Open file with append mode, without it concurrent fiber writes seem to conflict
+  string path = absl::StrCat(
+      base_path, "-", absl::Dec(ProactorBase::me()->GetPoolIndex(), absl::kZeroPad3), ".bin");
+  auto file = util::fb2::OpenWrite(path, io::WriteFile::Options{/*.append = */ false});
+  if (!file) {
+    LOG(ERROR) << "Error opening a file " << path << " for traffic logging: " << file.error();
+    return;
+  }
+  tl_traffic_logger.log_file = unique_ptr<io::WriteFile>{file.value()};
+}
+
+void LogTraffic(uint32_t id, bool has_more, absl::Span<RespExpr> resp) {
+  string_view cmd = resp.front().GetView();
+  if (absl::EqualsIgnoreCase(cmd, "debug"sv))
+    return;
+
+  DVLOG(2) << "Recording " << cmd;
+
+  char stack_buf[1024];
+  char* next = stack_buf;
+
+  // We write id, timestamp, has_more, num_parts, part_len, part_len, part_len, ...
+  // And then all the part blobs concatenated together.
+  auto write_u32 = [&next](uint32_t i) {
+    absl::little_endian::Store32(next, i);
+    next += 4;
+  };
+
+  write_u32(id);
+
+  absl::little_endian::Store64(next, absl::GetCurrentTimeNanos());
+  next += 8;
+
+  write_u32(has_more ? 1 : 0);
+  write_u32(uint32_t(resp.size()));
+
+  // Grab the lock and check if the file is still open.
+  lock_guard lk{tl_traffic_logger.mutex};
+
+  if (!tl_traffic_logger.log_file)
+    return;
+
+  // Proceed with writing the blob lengths.
+  for (auto part : resp) {
+    if (size_t(next - stack_buf + 4) > sizeof(stack_buf)) {
+      if (!tl_traffic_logger.Write(string_view{stack_buf, size_t(next - stack_buf)})) {
+        return;
+      }
+      next = stack_buf;
+    }
+    write_u32(part.GetView().size());
+  }
+
+  // Write the data itself.
+  std::array<iovec, 16> blobs;
+  unsigned index = 0;
+  if (next != stack_buf) {
+    blobs[index++] = iovec{.iov_base = stack_buf, .iov_len = size_t(next - stack_buf)};
+  }
+
+  for (auto part : resp) {
+    blobs[index++] = iovec{.iov_base = const_cast<char*>(part.GetView().data()),
```

Comment:
I'd at least safeguard somehow against larger commands

Reply:
why? it's not that i allocate memory here.

Reply:
Yes, that's why 🙂 so it doesn't go out of bounds